src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -94,59 +106,213 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
+            enrolledStudents, Array.from(studentEnrollRequests.values()));
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
+        });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
+    });
+  }
+
+  private partitionStudentEnrollRequests(studentEnrollRequests: StudentEnrollRequest[]): void {
+    let currentStudentChunk: StudentEnrollRequest[] = [];
+    for (const request of studentEnrollRequests) {
+      currentStudentChunk.push(request);
+      if (currentStudentChunk.length >= this.numberOfStudentsPerRequest) {
+        this.allStudentChunks.push(currentStudentChunk);
+        currentStudentChunk = [];
+      }
+    }
+    if (currentStudentChunk.length > 0) {
+      this.allStudentChunks.push(currentStudentChunk);
+    }
+  }
+
+  private checkTeamsValid(studentEnrollRequests: Map<number, StudentEnrollRequest>): void {
+    const teamSectionMap: Map<string, string> = new Map();
+    const teamIndexMap: Map<string, number> = new Map();
+    const noInvalidRowsBefore: number = this.invalidRowsIndex.size;
+
+    Array.from(studentEnrollRequests.keys()).forEach((key: number) => {
+      const request: StudentEnrollRequest | undefined = studentEnrollRequests.get(key);
+      if (request === undefined) {
+        return;
+      }
+
+      if (!teamSectionMap.has(request.team)) {
+        teamSectionMap.set(request.team, request.section);
+        teamIndexMap.set(request.team, key);
+        return;
+      }
+
+      if (teamSectionMap.get(request.team) !== request.section) {
+        this.invalidRowsIndex.add(key);
+        const firstIndex: number | undefined = teamIndexMap.get(request.team);
+        if (firstIndex !== undefined) {
+          this.invalidRowsIndex.add(firstIndex);
         }
-        this.isExistingStudentsPresent = true;
-      });
+      }
     });
+    if (this.invalidRowsIndex.size > noInvalidRowsBefore) {
+      this.enrollErrorMessage += 'Found duplicated teams in different sections. ';
+    }
+  }
+
+  private checkCompulsoryFields(studentEnrollRequests: Map<number, StudentEnrollRequest>): void {
+    const noInvalidRowsBefore: number = this.invalidRowsIndex.size;
+
+    Array.from(studentEnrollRequests.keys()).forEach((key: number) => {
+      const request: StudentEnrollRequest | undefined = studentEnrollRequests.get(key);
+      if (request === undefined) {
+        return;
+      }
+
+      if ((studentEnrollRequests.size >= 100 && request.section === '')
```

Comment:
you can just return early if size is smaller instead of entering this for loop. 

Also just put all the compulsory fields in an array first and check if any field in array is empty. (instead of doing many ors like this)

Reply:
I dont think we can return early if size is smaller than 100, since other fields apart from `section` could be empty
If we were to put fields in an array, `section` would an exception since it needs the bigger than 100 condition to be false

Reply:
ah fair enough - realised your doing multiple checks here